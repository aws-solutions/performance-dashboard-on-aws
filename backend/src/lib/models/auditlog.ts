/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
