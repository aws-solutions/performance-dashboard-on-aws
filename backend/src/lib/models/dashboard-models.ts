export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};