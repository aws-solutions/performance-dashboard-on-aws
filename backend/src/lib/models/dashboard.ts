/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { Widget } from "./widget";

export const DASHBOARD_ITEM_TYPE = "Dashboard";

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    /**
     * @type {string}
     * @deprecated in favor of {@link Draft}
     */
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export interface DashboardVersion {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
}

export interface Dashboard {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: Date;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: DashboardState;
    releaseNotes?: string;
    widgets?: Array<Widget>;
    friendlyURL?: string;
}

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string;
    sk: string;
    type: string;
    version: number;
    parentDashboardId: string;
    dashboardName: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: string;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: string;
    releaseNotes?: string;
    friendlyURL?: string;
}

// Public representation of a dashboard. Hides some fields
// like createdBy and state, which should not be exposed
// on public endpoints.
export interface PublicDashboard {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    updatedAt: Date;
    widgets?: Array<Widget>;
    friendlyURL?: string;

    // Filled on a homepage search with dashboard content
    // that matches the search query.
    queryMatches?: Array<string>;
}
