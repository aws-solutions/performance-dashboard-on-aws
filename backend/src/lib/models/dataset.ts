/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export interface Dataset {
    id: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: Date;
    sourceType: SourceType;
    schema?: DatasetSchema;
}

export type DatasetContent = Array<object>;
export type DatasetList = Array<Dataset>;

export interface DatasetItem {
    pk: string;
    sk: string;
    type: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: string;
    sourceType: string;
    schema?: string;
}
