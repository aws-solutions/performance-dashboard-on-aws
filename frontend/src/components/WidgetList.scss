@import "../styles/base";

.text-overflow-ellipsis {
  text-overflow: ellipsis;
}

.minus-2-margin {
  margin-right: -0.5em;
}
