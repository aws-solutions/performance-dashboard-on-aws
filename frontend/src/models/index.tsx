/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export type TopicArea = {
    id: string;
    name: string;
    createdBy: string;
    dashboardCount: number;
};

export const TopicAreaSortingCriteria = (a: TopicArea, b: TopicArea) => (a.name > b.name ? 1 : -1);

export type PublicTopicArea = {
    id: string;
    name: string;
    dashboards?: Array<Dashboard | PublicDashboard>;
};

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetType {
    DynamicDataset = "DynamicDataset",
    StaticDataset = "StaticDataset",
    CsvFileUpload = "CsvFileUpload",
    CreateNew = "CreateNew",
}

export type Dashboard = {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    releaseNotes?: string;
    widgets: Array<Widget>;
    state: string;
    updatedAt: Date;
    createdBy: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type PublicDashboard = {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description?: string;
    widgets: Array<Widget>;
    updatedAt: Date;
    friendlyURL?: string;
    queryMatches?: Array<string>;
};

export type DashboardVersion = {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
};

export type DashboardAuditLog = {
    itemId: string;
    timestamp: Date;
    version: number;
    userId: string;
    event: "Create" | "Update" | "Delete";
    modifiedProperties?: Array<{
        property: string;
        oldValue: any;
        newValue: any;
    }>;
};

export enum WidgetType {
    Text = "Text",
    Chart = "Chart",
    Table = "Table",
    Image = "Image",
    Metrics = "Metrics",
    Section = "Section",
}

export enum ChartType {
    LineChart = "LineChart",
    ColumnChart = "ColumnChart",
    BarChart = "BarChart",
    PartWholeChart = "PartWholeChart",
    PieChart = "PieChart",
    DonutChart = "DonutChart",
}

export interface Widget {
    id: string;
    name: string;
    widgetType: string;
    order: number;
    updatedAt: Date;
    dashboardId: string;
    content: any;
    showTitle: boolean;
    section?: string;
}

export interface ChartWidget extends Widget {
    content: {
        title: string;
        chartType: ChartType;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        horizontalScroll?: boolean;
        stackedChart?: boolean;
        significantDigitLabels: boolean;
        dataLabels: boolean;
        computePercentages: boolean;
        showTotal: boolean;
    };
}

export interface TableWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        summary: string;
        summaryBelow: boolean;
        datasetType?: DatasetType;
        s3Key: {
            raw: string;
            json: string;
        };
        columnsMetadata: ColumnMetadata[];
        sortByColumn?: string;
        sortByDesc?: boolean;
        significantDigitLabels: boolean;
        displayWithPages: boolean;
    };
}

export interface ImageWidget extends Widget {
    content: {
        title: string;
        imageAltText: string;
        summary?: string;
        summaryBelow: boolean;
        s3Key: {
            raw: string;
        };
        filename: string;
        scalePct?: string;
    };
}

export interface MetricsWidget extends Widget {
    content: {
        title: string;
        datasetId: string;
        oneMetricPerRow: boolean;
        datasetType?: DatasetType;
        significantDigitLabels: boolean;
        metricsCenterAlign: boolean;
        s3Key: {
            raw: string;
            json: string;
        };
    };
}

export interface SectionWidget extends Widget {
    content: {
        title: string;
        summary: string;
        widgetIds?: Array<string>;
        showWithTabs: boolean;
        horizontally?: boolean;
    };
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export type Dataset = {
    id: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    createdBy?: string;
    updatedAt?: Date;
    sourceType?: SourceType;
    schema?: DatasetSchema;
};

export type PublicHomepage = {
    title: string;
    description: string;
    dashboards: Array<PublicDashboard>;
};

export type Homepage = {
    title: string;
    description: string;
    updatedAt: Date;
};

export type PublicSettings = {
    dateTimeFormat: {
        date: string;
        time: string;
    };
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels?: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Settings = {
    publishingGuidance: string;
    dateTimeFormat: {
        date: string;
        time: string;
    };
    updatedAt?: Date;
    navbarTitle?: string;
    colors: {
        primary: string;
        secondary: string;
    };
    topicAreaLabels: {
        singular: string;
        plural: string;
    };
    customLogoS3Key?: string;
    customLogoAltText?: string;
    customFaviconS3Key?: string;
    contactEmailAddress?: string;
    adminContactEmailAddress?: string;
    contactUsContent?: string;
    analyticsTrackingId?: string;
};

export type Metric = {
    title: string;
    value: number;
    changeOverTime?: string;
    percentage?: string;
    currency?: string;
    startDate?: string;
    endDate?: string;
};

// Type for the History object in react-router. Defines the
// location state that is common across all routes.
// Usage in a component: useHistory<LocationState>();
export type LocationState = {
    alert?: {
        type: "success" | "warning" | "info" | "error";
        message: string;
        to?: string;
        linkLabel?: string;
    };
    emails?: string;
    usernames?: Array<string>;
    id?: string;
    metrics?: Array<Metric>;
    metric?: Metric;
    position?: number;
    showTitle?: boolean;
    oneMetricPerRow?: boolean;
    significantDigitLabels?: boolean;
    metricsCenterAlign?: boolean;
    metricTitle?: string;
    origin?: string;
    json?: Array<any>;
    staticDataset?: Dataset;
    redirectUrl?: string;
    crumbLabel?: string;
    datasetType?: DatasetType;
};

export enum UserRoles {
    Admin = "Admin",
    Editor = "Editor",
    /**
     * @type {string}
     * @deprecated no longer supported
     */
    Publisher = "Publisher",
    Public = "Public",
}

export type User = {
    userId: string;
    enabled: boolean;
    userStatus: string;
    sub: string;
    email: string;
    roles: Array<UserRoles>;
    createdAt: Date;
    updatedAt: Date;
};

export enum ColumnDataType {
    Number = "Number",
    Text = "Text",
    Date = "Date",
}

export enum NumberDataType {
    Percentage = "Percentage",
    Currency = "Currency",
    "With thousands separators" = "With thousands separators",
}

export enum CurrencyDataType {
    "Dollar $" = "Dollar $",
    "Euro €" = "Euro €",
    "Pound £" = "Pound £",
}

export interface ColumnMetadata {
    columnName: string;
    dataType?: ColumnDataType;
    numberType?: NumberDataType;
    currencyType?: CurrencyDataType;
    hidden: boolean;
}
