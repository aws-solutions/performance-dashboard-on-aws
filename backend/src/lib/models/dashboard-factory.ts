import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
