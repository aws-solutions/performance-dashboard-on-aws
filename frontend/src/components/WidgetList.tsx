import React, { createRef } from "react";
import { Widget, WidgetType } from "../models";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faArrowUp, faArrowDown } from "@fortawesome/free-solid-svg-icons";
import Button from "./Button";
import "./WidgetList.css";
import Link from "./Link";
import AlertContainer from "../containers/AlertContainer";
import SecondaryActionBar from "./SecondaryActionBar";
import ContentItem from "./ContentItem";
import { useTranslation } from "react-i18next";

interface Props {
  onClick: Function;
  onDelete?: Function;
  onDuplicate?: Function;
  onMoveUp?: Function;
  onMoveDown?: Function;
  widgets: Array<Widget>;
}

function WidgetList(props: Props) {
  const { t } = useTranslation();
  const caretUpRefs = props.widgets.map(() => createRef<HTMLButtonElement>());
  const caretDownRefs = props.widgets.map(() => createRef<HTMLButtonElement>());

  const onDelete = (widget: Widget) => {
    if (props.onDelete) {
      props.onDelete(widget);
    }
  };

  const onDuplicate = (widget: Widget) => {
    if (props.onDuplicate) {
      props.onDuplicate(widget);
    }
  };

  const onMoveDown = (index: number) => {
    if (props.onMoveDown) {
      setNextFocus(index, "Down");
      props.onMoveDown(index);
    }
  };

  const onMoveUp = (index: number) => {
    if (props.onMoveUp) {
      setNextFocus(index, "Up");
      props.onMoveUp(index);
    }
  };

  const setNextFocus = (index: number, direction: "Up" | "Down") => {
    let ref;
    /**
     * When moving a widget up or down, the focus should follow the
     * widget to its new position in the list. Determining the new
     * position depends on whether the widget is moving up or down
     * and if the widget has reached a boundary (i.e. beginning or
     * end of the list), in which case the focus should swap to the
     * opposite caret (careUp vs caretDown).
     *
     * We may not need this logic when we implement drag-n-drop
     * because most of the existing libraries already handle browser
     * focus when moving DOM objects.
     */
    if (direction === "Up") {
      const next = index - 1;
      ref = next === 0 ? caretDownRefs[0] : caretUpRefs[next];
    } else {
      const next = index + 1;
      const last = props.widgets.length - 1;
      ref = next === last ? caretUpRefs[last] : caretDownRefs[next];
    }

    if (ref.current) {
      ref.current.focus();
    }
  };

  return (
    <div>
      {props.widgets && props.widgets.length ? (
        <div>
          <AlertContainer />
          <SecondaryActionBar stickyPosition={160}>
            <h3 className="margin-bottom-0 margin-top-0">
              {t("DashboardContent")}
            </h3>
            <p className="margin-top-2px margin-bottom-0">
              {t("BuildDashboardGuidance")}
            </p>
            <div className="grid-row radius-lg margin-top-4 text-bold font-sans-sm">
              <div className="grid-col flex-1">
                <div className="margin-left-1">{t("Order")}</div>
              </div>
              <div className="grid-col flex-5">
                <div className="margin-left-2">{t("NameUpperCase")}</div>
              </div>
              <div className="grid-col flex-6">
                <div className="margin-left-6">{t("ContentType")}</div>
              </div>
            </div>
          </SecondaryActionBar>
          {props.widgets.map((widget, index) => {
            return (
              <ContentItem className="grid-row margin-y-1" key={index}>
                <div className="grid-row grid-col flex-1 padding-1">
                  <div className="grid-col flex-6 text-center display-flex flex-align-center flex-justify-center font-sans-md">
                    {index + 1}
                  </div>
                  <div className="grid-col flex-6 grid-row flex-column text-center">
                    <div className="grid-col flex-6">
                      {index > 0 && (
                        <Button
                          variant="unstyled"
                          className="text-base-darker hover:text-base-darkest active:text-base-darkest"
                          ariaLabel={t("MoveContentItemUp", {
                            name: widget.name,
                          })}
                          onClick={() => onMoveUp(index)}
                          ref={caretUpRefs[index]}
                        >
                          <FontAwesomeIcon size="sm" icon={faArrowUp} />
                        </Button>
                      )}
                    </div>
                    <div className="grid-col flex-6">
                      {index < props.widgets.length - 1 && (
                        <Button
                          variant="unstyled"
                          className="text-base-darker hover:text-base-darkest active:text-base-darkest"
                          ariaLabel={t("MoveContentItemDown", {
                            name: widget.name,
                          })}
                          onClick={() => onMoveDown(index)}
                          ref={caretDownRefs[index]}
                        >
                          <FontAwesomeIcon
                            id={`${widget.id}-move-down`}
                            size="sm"
                            icon={faArrowDown}
                          />
                        </Button>
                      )}
                    </div>
                  </div>
                </div>
                <div className="border-base-lighter border-left"></div>
                <div className="grid-col flex-11 grid-row padding-1 margin-y-1">
                  <div
                    className="grid-col flex-6 usa-tooltip text-bold"
                    data-position="bottom"
                    title={widget.name}
                  >
                    <div className="margin-left-1 text-no-wrap overflow-hidden text-overflow-ellipsis">
                      {widget.name}
                    </div>
                  </div>
                  <div className="grid-col flex-3 text-italic">
                    {t(
                      widget.widgetType === WidgetType.Chart
                        ? widget.content.chartType
                        : widget.widgetType
                    )}
                  </div>
                  <div className="grid-col flex-3 text-right">
                    <Link
                      ariaLabel={t("EditContent", { name: widget.name })}
                      to={`/admin/dashboard/${
                        widget.dashboardId
                      }/edit-${widget.widgetType.toLowerCase()}/${widget.id}`}
                    >
                      {t("Edit")}
                    </Link>
                    <Button
                      variant="unstyled"
                      className="margin-left-2 text-base-dark hover:text-base-darker active:text-base-darkest"
                      onClick={() => onDuplicate(widget)}
                      ariaLabel={t("DuplicateContent", { name: widget.name })}
                    >
                      {t("Duplicate")}
                    </Button>
                    <Button
                      variant="unstyled"
                      className="margin-left-2 text-base-dark hover:text-base-darker active:text-base-darkest"
                      onClick={() => onDelete(widget)}
                      ariaLabel={t("DeleteContent", { name: widget.name })}
                    >
                      {t("Delete")}
                    </Button>
                  </div>
                </div>
              </ContentItem>
            );
          })}
          <div className="text-center margin-top-2">
            <button
              className="usa-button usa-button--base margin-top-1"
              onClick={() => {
                if (props.onClick) {
                  props.onClick();
                }
              }}
            >
              {t("PlusAddContentItem")}
            </button>
          </div>
        </div>
      ) : (
        <SecondaryActionBar className="text-center padding-5 margin-y-2">
          <div>
            <p>
              {t("NoContentItems")} <br />
              {t("ChartsTablesMore")}
            </p>
            <div className="text-center margin-top-4">
              <Button
                className="margin-top-1"
                variant="base"
                onClick={() => {
                  if (props.onClick) {
                    props.onClick();
                  }
                }}
              >
                {t("PlusAddContentItem")}
              </Button>
            </div>
          </div>
        </SecondaryActionBar>
      )}
    </div>
  );
}

export default WidgetList;
