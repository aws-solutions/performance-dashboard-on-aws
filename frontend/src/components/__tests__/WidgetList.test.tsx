import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import { Widget } from "../../models";
import WidgetList from "../WidgetList";

const widgets: Array<Widget> = [
  {
    id: "123",
    name: "The benefits of bananas",
    widgetType: "Text",
    order: 1,
    updatedAt: "",
  },
];

test("renders an empty box", async () => {
  const onClick = jest.fn();
  const wrapper = render(<WidgetList widgets={[]} onClick={onClick} />);
  expect(wrapper.container).toMatchSnapshot();
});

test("calls the onClick callback when button is pressed", async () => {
  const onClick = jest.fn();
  const wrapper = render(<WidgetList widgets={[]} onClick={onClick} />);
  fireEvent.click(wrapper.getByRole("button"));
  expect(onClick).toBeCalled();
});

test("calls onDelete when user clicks delete button", async () => {
  const onDelete = jest.fn();
  const onClick = jest.fn();
  const wrapper = render(
    <WidgetList widgets={widgets} onClick={onClick} onDelete={onDelete} />
  );

  await act(async () => {
    fireEvent.click(wrapper.getByRole("button", { name: "Delete" }));
  });

  expect(onDelete).toBeCalled();
});

test("calls onMoveUp when user clicks up arrow", async () => {
  const onMoveUp = jest.fn();
  const onClick = jest.fn();
  const wrapper = render(
    <WidgetList widgets={widgets} onClick={onClick} onMoveUp={onMoveUp} />
  );

  await act(async () => {
    fireEvent.click(
      wrapper.getByRole("button", {
        name: "Move The benefits of bananas up",
      })
    );
  });

  expect(onMoveUp).toBeCalled();
});

test("calls onMoveUp when user clicks down arrow", async () => {
  const onMoveDown = jest.fn();
  const onClick = jest.fn();
  const wrapper = render(
    <WidgetList widgets={widgets} onClick={onClick} onMoveDown={onMoveDown} />
  );

  await act(async () => {
    fireEvent.click(
      wrapper.getByRole("button", {
        name: "Move The benefits of bananas down",
      })
    );
  });

  expect(onMoveDown).toBeCalled();
});
