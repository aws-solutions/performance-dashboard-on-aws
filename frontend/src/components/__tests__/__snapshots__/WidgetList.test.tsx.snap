// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an empty box 1`] = `
<div>
  <div>
    <div
      class="text-center radius-lg padding-5 margin-y-3 border-base border-dashed bg-base-lightest border"
    >
      <p>
        This dashboard is empty. Build the dashboard by adding 
        <br />
        charts, tables, and text as content.
      </p>
      <div
        class="text-center margin-top-4"
      >
        <button
          class="usa-button usa-button--base margin-top-1"
        >
          + Add content
        </button>
      </div>
    </div>
  </div>
</div>
`;
